backend/src/datasources/postDatasource.js

```diff
@@ -0,0 +1,106 @@
+import { DataSource } from 'apollo-datasource'
+import { UserInputError } from 'apollo-server'
+import { Post } from '../models/post'
+
+export class PostDataSource extends DataSource {
+  constructor (posts = []) {
+    super()
+    this.posts = posts
+  }
+
+  initialize ({ context }) {
+    this.context = context
+  }
+
+  async getPostById (id) {
+    if (id) {
+      return this.posts.find((post) => post.id === id)
+    } else {
+      throw new UserInputError(
+        'No matching post found. Please check your input'
+      )
+    }
+  }
+
+  async getPostByTitle (title) {
+    if (title) {
+      return this.posts.find((post) => post.title === title)
+    } else {
+      throw new UserInputError(
+        'No matching post found. Please check your input'
+      )
+    }
+  }
+
+  async deletePost (args) {
+    const postFound = this.posts.find((post) => post.id === args.id)
+    if (postFound) {
+      this.posts = this.posts.filter((post) => {
```

Comment:
Consider the following suggestion.
```suggestion
      this.posts = this.posts.filter((post) => post.id !== postFound.id)
```